Modules/MUON/MID/src/MIDTrending.cxx

```diff
@@ -0,0 +1,177 @@
+#include "QualityControl/ObjectMetadataKeys.h"
+#include "QualityControl/RootClassFactory.h"
+#include "QualityControl/RepoPathUtils.h"
+#include <TDatime.h>
+#include <TH1.h>
+#include <TCanvas.h>
+#include <TPaveText.h>
+#include <TGraphErrors.h>
+#include <TPoint.h>
+#include <map>
+#include <string>
+#include <memory>
+#include <vector>
+using namespace o2::quality_control;
+using namespace o2::quality_control::core;
+using namespace o2::quality_control::repository;
+using namespace o2::quality_control::postprocessing;
+
+void MIDTrending::configure(const boost::property_tree::ptree& config)
+{
+  mConfig = TrendingTaskConfigMID(getID(), config);
+}
+
+void MIDTrending::initialize(Trigger, framework::ServiceRegistryRef services)
+{
+  // Preparing data structure of TTree
+  mTrend = std::make_unique<TTree>();
+  mTrend->SetName(PostProcessingInterface::getName().c_str());
+  mTrend->Branch("runNumber", &mMetaData.runNumber);
+
+  mTrend->Branch("ntreeentries", &ntreeentries);
+  mTrend->Branch("time", &mTime);
+
+  for (const auto& source : mConfig.dataSources) {
+    std::unique_ptr<Reductor> reductor(root_class_factory::create<Reductor>(source.moduleName, source.reductorName));
+
+    mTrend->Branch(source.name.c_str(), reductor->getBranchAddress(), reductor->getBranchLeafList());
+    mReductors[source.name] = std::move(reductor);
+  }
+}
+
+// todo: see if OptimizeBaskets() indeed helps after some time
+void MIDTrending::update(Trigger t, framework::ServiceRegistryRef services)
+{
+  auto& qcdb = services.get<repository::DatabaseInterface>();
+  trendValues(t, qcdb);
+  generatePlots(qcdb);
+}
+
+void MIDTrending::finalize(Trigger, framework::ServiceRegistryRef services)
+{
+
+  auto& qcdb = services.get<repository::DatabaseInterface>();
+  auto mo = std::make_shared<core::MonitorObject>(mTrend.get(), getName(), "o2::quality_control_modules::trd::MIDTrending", mConfig.detectorName);
+  mo->setIsOwner(false);
+  qcdb.storeMO(mo);
+  generatePlots(qcdb);
+}
+
+void MIDTrending::trendValues(const Trigger& t, repository::DatabaseInterface& qcdb)
+{
+  mTime = TDatime().Convert(); // ROOT expects seconds since epoch.
+  mMetaData.runNumber = t.activity.mId;
+  int count = 0;
+
+  for (auto& dataSource : mConfig.dataSources) {
+
+    // todo: make it agnostic to MOs, QOs or other objects. Let the reductor cast to whatever it needs.
+    if (dataSource.type == "repository") {
+      auto mo = qcdb.retrieveMO(dataSource.path, dataSource.name, t.timestamp);
+      if (mo == nullptr) {
+        ILOG(Warning, Devel) << "Could not retrieve MO '" << dataSource.name << "' from QCDB, skipping this data source" << ENDM;
+        continue;
+      }
+
+      if (!count) {
+        std::map<std::string, std::string> entryMetadata = mo->getMetadataMap();  // full list of metadata as a map
+        mMetaData.runNumber = std::stoi(entryMetadata[metadata_keys::runNumber]); // get and set run number
+        ntreeentries = (Int_t)mTrend->GetEntries() + 1;
```

Comment:
Out of curiosity, why do you keep track the number of entries in the tree and store it there, if TTree has `GetEntries`?

Reply:
I did not change this part from the template.